Knoema.Client/Client.cs

```diff
@@ -18,343 +19,540 @@
+
+			if (!string.IsNullOrEmpty(_clientId) && !string.IsNullOrEmpty(_clientSecret))
+				message.Headers.Add("Authorization", GetAuthorizationHeaderValue(_clientId, _clientSecret));
+
+			var sendAsyncResp = await GetApiClient().SendAsync(message);
+			sendAsyncResp.EnsureSuccessStatusCode();
+			var strRead = await sendAsyncResp.Content.ReadAsStringAsync();
+			return JsonConvert.DeserializeObject<SearchTimeSeriesResponse>(strRead);
+		}
+
+		private static string GetAuthorizationHeaderValue(string clientId, string clientSecret)
+		{
+			return string.Format("Knoema {0}:{1}:{2}", clientId,
+					Convert.ToBase64String(new HMACSHA1(Encoding.UTF8.GetBytes(DateTime.UtcNow.ToString("dd-MM-yy-HH"))).ComputeHash(Encoding.UTF8.GetBytes(clientSecret))),
+					AuthProtoVersion);
+		}
+
+		private static Uri GetUri(string host, string token, string path, Dictionary<string, string> parameters = null)
+		{
+			if (!string.IsNullOrEmpty(token))
+			{
+				if (parameters == null)
+					parameters = new Dictionary<string, string>();
+				parameters.Add("access_token", token);
+			}
+			var builder = new UriBuilder(Uri.UriSchemeHttp, host)
+			{
+				Path = path,
+				Query = parameters != null ?
+					string.Join("&", parameters.Select(pair => string.Format("{0}={1}", pair.Key, HttpUtility.UrlEncode(pair.Value)))) :
+					string.Empty
+			};
+			return builder.Uri;
+		}
+
+		public Task<T> GetTaskResult<T>(int taskKey) where T : TaskResult
+		{
+			return ApiGet<T>("/api/1.0/meta/taskresult", string.Format("taskKey={0}", taskKey));
+		}
+
+		public async Task<T> WaitTaskResult<T>(int taskKey, int spinDelayInSeconds, int maxWaitCount) where T : TaskResult
+		{
+			T taskResult = null;
+			for (var i = 0; ;)
+			{
+				taskResult = await GetTaskResult<T>(taskKey);
+				if (!(taskResult.Status == Meta.TaskStatus.Executing || taskResult.Status == Meta.TaskStatus.Pending))
+					break;
+				i++;
+				if (i >= maxWaitCount)
+					throw new Exception("Maximum wait count reached");
+				Thread.Sleep(spinDelayInSeconds * 1000);
+			}
+
+			if (taskResult.Status == Meta.TaskStatus.Cancelled)
+				throw new TaskCanceledException("Task was cancelled");
+			if (taskResult.Status == Meta.TaskStatus.Failed)
+				throw new Exception(taskResult.Message);
+
+			if (taskResult.Status == Meta.TaskStatus.Completed)
+				return taskResult;
+
+			throw new ArgumentOutOfRangeException("Unexpected task status");
+		}
+
+		public Task<UnloadResponse> Unload(PivotRequest request)
+		{
+			return ApiPost<UnloadResponse>("/api/1.0/data/unload", request);
+		}
+
+		public async Task<DatasetUnloadTaskResultData> UnloadData(PivotRequest request, int spinDelayInSeconds = 10, int maxWaitCount = 360)
+		{
+			var unloadTask = await Unload(request);
+			var taskResult = await WaitTaskResult<DatasetUnloadTaskResult>(unloadTask.TaskKey, spinDelayInSeconds, maxWaitCount);
+			return taskResult.Data;
+		}
+
+		public Task<Task>[] UnloadGetFiles(string[] files, Stream[] resultStreams)
```

Comment:
Looks like an internal private methoud as well.
Strange name by the way.

Reply:
ok, renamed